### ISSUE TITLE
Runtime Exception - Failed to read, graphDef and prefix cannot be null

### ISSUE THREAD

**Issue by dreistheman** [szaza/tensorflow-example-java#6]
@szaza I'm trying to run your Java API example (https://sites.google.com/view/tensorflow-example-java-api/home?authuser=0). 
-Downloaded the graph and label file from the provided gdrive link, placed it in _tensorflow-example-java\src\main\resources\YOLO_
-edited the configuration file with the absolute paths of the yolo-voc.pb and yolo-voc-labels.pb

Unfortunately resulted to ServiceException and IllegalArgumentException. I'm trying to test this out so I can use it with my own trained model (at 20,000 steps now) following your tutorial too (Train YOLO with DarkFlow) 

I am using Java JDK 1.8 and TensorFlow 1.9. Thank you in advance. :D

**Comment by szaza** [comment]
Hi,
Could you please copy the whole stack trace? It would help me a lot to see where the exception is thrown. Could you try to use relative paths in your configuration file? They should be relative from the _/resources/YOLO_ directory. Please also try to start the server in debug mode and put brake points before the line where the exception is thrown. Run the application line by line from that point and hopefully you'll see the problem.

**Comment by dreistheman** [comment]
Hello, thank you for replying. :) I now changed the paths to relative from the Config.java's directory. Here's the stack trace:
00:51:19.004 [main] ERROR edu.ml.tensorflow.util.IOUtil - Failed to read [../../../../resources/YOLO/yolo-voc.pb]!
00:51:19.004 [main] ERROR edu.ml.tensorflow.ObjectDetector - Download one of my graph file to run the program!
You can find my graphs here: https://drive.google.com/open?id=1GfS1Yle7Xari1tRUEi2EDYedFteAOaoN
2019-04-09 00:51:21.548448: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2
Exception in thread "main" java.lang.IllegalArgumentException: graphDef and prefix cannot be null
        at org.tensorflow.Graph.importGraphDef(Graph.java:127)
        at org.tensorflow.Graph.importGraphDef(Graph.java:114)
        at edu.ml.tensorflow.ObjectDetector.executeYOLOGraph(ObjectDetector.java:89)
        at edu.ml.tensorflow.ObjectDetector.detect(ObjectDetector.java:49)
        at edu.ml.tensorflow.Main.main(Main.java:8)

**Comment by szaza** [comment]
I can see the problem, your paths are wrong.
`String GRAPH_FILE = "../../../../resources/YOLO/yolo-voc.pb";` - it means that you have near the project root a /resources/YOLO directory with yolo-voc.pb.
I would recommend just to use the default values and place your .pb and txt file into the _/src/main/resources_ folder. I would also recommend to use the _./sample_ value as OUTPUT_DIR;

tensorflow-example-java/
 - src/
   - main/
      - resources/
          -> **your .pb and .txt files should be placed here**

Please read more detail about the maven project structure here: https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html; Gradle uses the same archetypes.

**Comment by dreistheman** [comment]
It worked now! Yes I actually read back again on your first reply mentioning "They should be relative from the /resources/YOLO directory."  after just posting my recent response. My lack of experience assumed I would use the same convention of "navigating from the current directory". 

Last quick small side question: I'm training 4 classes now at 20,000 steps on top of tiny-yolo-voc. Would it be better to stop at 30 or 40,000 for 4 classes?

**Comment by szaza** [comment]
It can happen, however it also depends on your training set and your settings. I would suggest to save your models after every 10,000 steps. When you discover that the confidence rate doesn't grow higher (it is constant) or it is lower than the steps before, than you definitely have to stop the training process.
I'm not sure whether you are familiar with Courseras Machine Learning courses: https://www.coursera.org/learn/machine-learning?
If your're not I would suggest to go through on the lessons hold by Andrew Ng. I've got many-many good answers related with my ML related questions. 
Best regards, and good luck for your work!

**Comment by dreistheman** [comment]
Didn't change anything on the tiny-yolo-voc-4c.cfg except the "classes" and "filters" attribute. Is the confidence rate the "loss" or "moving ave loss", which in my case is playing around 0.1-0.4 value for a while now I noticed. Steps, loss, and moving ave loss are the only data displayed in the console training on darkflow. Also, how do I save the current model while it's training? 

I'm not familiar with that yet so I'll definitely have the course. Thank you for your outstanding help in all these. :)
